### PR TITLE
#1654 GPU Index Flat Delete

### LINKED ISSUE #1654: GPU IndexFlat support deleted vectors searching.
**Is your feature request related to a problem? Please describe.**
In 0.7.0, No GPU index support deleted vectors searching.

**Describe the solution you'd like**
Similar with CPU index, just use the bitmap to filt out the deleted vector in searching phase.

### PR DESCRIPTION
**Pull request by sahuang**
**What type of PR is this?**

feature

**What this PR does / why we need it:**

Support Index Flat Deletion

**Which issue(s) this PR fixes:**

Fixes #1654 

**Special notes for your reviewer:**

**Does this PR introduce a user-facing change?:**

**Additional documentation (e.g. design docs, usage docs, etc.):**

### CODE REVIEW

File: core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh [hunk 1]
```diff
@@ -56,13 +56,15 @@ class FlatIndex {
...
+             bool exactDistance,
+             Tensor<uint8_t, 1, true> bitset);
 
   void query(Tensor<half, 2, true>& vecs,
              int k,
              Tensor<half, 2, true>& outDistances,
              Tensor<int, 2, true>& outIndices,
-             bool exactDistance);
+             bool exactDistance,
+             Tensor<uint8_t, 1, true> bitset);
```
Comment (cydrain) [inline core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh]: move before "int k"

File: core/src/index/thirdparty/faiss/gpu/impl/IVFFlat.cu [hunk 2]
```diff
@@ -329,7 +330,8 @@ IVFFlat::query(Tensor<float, 2, true>& queries,
                int nprobe,
                int k,
                Tensor<float, 2, true>& outDistances,
-               Tensor<long, 2, true>& outIndices) {
+               Tensor<long, 2, true>& outIndices,
+               Tensor<uint8_t, 1, true> bitset) {
```
Comment (cydrain) [inline core/src/index/thirdparty/faiss/gpu/impl/IVFFlat.cu]: move before "int k"

File: core/src/index/thirdparty/faiss/gpu/impl/L2Select.cu [hunk 3]
```diff
@@ -159,13 +159,164 @@ __global__ void l2SelectMinK(Tensor<T, 2, true> productDistances,
...
+
 template <typename T>
 void runL2SelectMin(Tensor<T, 2, true>& productDistances,
                     Tensor<T, 1, true>& centroidDistances,
                     Tensor<T, 2, true>& outDistances,
                     Tensor<int, 2, true>& outIndices,
                     int k,
-                    cudaStream_t stream) {
+                    cudaStream_t stream,
+                    Tensor<uint8_t, 1, true> bitset) {
```
Comment (cydrain) [inline core/src/index/thirdparty/faiss/gpu/impl/L2Select.cu]: move before "outDistances"

File: core/src/index/thirdparty/faiss/gpu/utils/BlockSelectHalf.cu [hunk 4]
```diff
@@ -45,7 +45,8 @@ BLOCK_SELECT_DECL(half, false, 2048);
 void runBlockSelect(Tensor<half, 2, true>& in,
                     Tensor<half, 2, true>& outK,
                     Tensor<int, 2, true>& outV,
-                    bool dir, int k, cudaStream_t stream) {
+                    bool dir, int k, cudaStream_t stream,
+                    Tensor<uint8_t, 1, true> bitset) {
```
Comment (cydrain) [inline core/src/index/thirdparty/faiss/gpu/utils/BlockSelectHalf.cu]: move before "outK"

File: core/src/index/thirdparty/faiss/gpu/GpuIndexIVFScalarQuantizer.cu [hunk 5]
```diff
@@ -258,7 +258,8 @@ GpuIndexIVFScalarQuantizer::searchImpl_(int n,
                                         const float* x,
                                         int k,
                                         float* distances,
-                                        Index::idx_t* labels) const {
+                                        Index::idx_t* labels,
+                                        ConcurrentBitsetPtr bitset) const {
```
Comment (cydrain) [inline core/src/index/thirdparty/faiss/gpu/GpuIndexIVFScalarQuantizer.cu]: add default value "bitset = nullptr"

Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/GpuIndexIVFScalarQuantizer.cu]: Fixed all nullptr default arg

File: core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh (hunk 1 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh]: done

File: core/src/index/thirdparty/faiss/gpu/impl/IVFFlat.cu (hunk 2 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/impl/IVFFlat.cu]: ok

File: core/src/index/thirdparty/faiss/gpu/impl/L2Select.cu (hunk 3 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/impl/L2Select.cu]: done

File: core/src/index/thirdparty/faiss/gpu/utils/BlockSelectHalf.cu (hunk 4 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/utils/BlockSelectHalf.cu]: done

File: core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh (hunk 1 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh]: ok

File: core/src/index/thirdparty/faiss/gpu/impl/IVFFlat.cu (hunk 2 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/impl/IVFFlat.cu]: ok

File: core/src/index/thirdparty/faiss/gpu/impl/L2Select.cu (hunk 3 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/impl/L2Select.cu]: ok

File: core/src/index/thirdparty/faiss/gpu/GpuIndexIVFScalarQuantizer.cu (hunk 5 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/GpuIndexIVFScalarQuantizer.cu]: added

File: core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh [hunk 6]
```diff
@@ -53,12 +53,14 @@ class FlatIndex {
                                                      cudaStream_t stream);
 
   void query(Tensor<float, 2, true>& vecs,
+             Tensor<uint8_t, 1, true> bitset,
```
Comment (cydrain) [inline core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh]: Tensor<uint8_t, 1, true>& bitset,

**Review by wxyucs — APPROVED** [review]

File: core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh (hunk 6 above)
Comment (sahuang) [reply core/src/index/thirdparty/faiss/gpu/impl/FlatIndex.cuh]: added reference

**Pull request merged**